`Pagination` - "How to use" and "Component API" sections in documentation for Helios website (#1128)
hashicorp/design-system: website/docs/components/pagination/partials/code/how-to-use.md

## "Compact" vs "Numbered" pagination

There are two different variants of the `Pagination` component (with different ways to invoke them) built to cover different use cases, contexts, and designs you may need them for.

This differentiation is necessary to cover both use cases of a pagination for a list with a known number of elements (**"numbered"**) and one in which this information is not available or is [cursor-based](https://jsonapi.org/profiles/ethanresnick/cursor-pagination/) (**"compact"**).

In the first one, the user is presented with a list of navigation controls ("prev/next" and "page numbers" to go directly to a specific page) and other optional UI elements; in the second, much simpler one, the user is presented with only the "prev/next" controls.

When pagination is invoked directly using one of these two components, it will **automatically**:

- provide the correct responsive layout for the entire pagination and its sub-parts

> - manage the "current page" status across the different sub-components it's made of (based on the arguments provided to it and its children).

Reviewer: ```suggestion
- manage the "current page" status across the different sub-components it’s made of (based on the arguments provided to it and its children).
```

This means that if you need to update the URL when the user changes the "page" in the Pagination (eg. to add/remove/update some query parameters), you have to do it whithin the `onPageChange` callback you provide to the component.

If instead you need to update the URL directly when the user clicks on one of the "navigation control" elements, you have to provide routing parameters (`route/query/model/etc`) to the component; refer to the "Component API" section below for specifications about these parameters (the APIs are slightly different for the two components).

## How to use the "Numbered" pagination

The basic invocation of the "numbered" pagination requires the `@totalItems` argument to be provided (plus the event/routing handlers, see below):

```handlebars
<Hds::Pagination::Numbered @totalItems={{40}} />
```

By default the "Info" and "SizeSelector" are displayed, and the component takes care of updating the values and the states of the different elements, according to the user interactions with the component.

### Extra arguments

> It's possible to customize the "Info", "Controls", and "SizeSelector" providing additional arguments to them. For more details about these parameters, refer to the "Components API" section.

Reviewer: ```suggestion
It’s possible to customize the "Info", "Controls", and "SizeSelector" components by providing additional arguments to them. For more details about these parameters, refer to the "Component API" section.
```

Below is an example of some of these extra arguments:

```handlebars
<Hds::Pagination::Numbered
  @totalItems={{40}}
  @itemsPerPage={{20}}
  @showTotalItems={{false}}
  @showSizeSelector={{false}}
  @pageSizes={{array 5 20 60}}
/>
```

### Truncation

> When there is a large number of items and consequently the number of pages is also large, by default the component automatically "truncates" the number of visible pages (using "ellipsises"):

Reviewer: ```suggestion
When there is a large number of items and consequently the number of pages is also large, by default the component automatically "truncates" the number of visible pages (using "ellipses"):
```

Author: 🤦‍♂️

```handlebars
<Hds::Pagination::Numbered
  @totalItems={{120}}
/>
```

> If necessary, it's possible to disable this functionality using a specific `@isTruncated` argument:

Reviewer: ```suggestion
If necessary, it’s possible to disable this functionality using the `@isTruncated` argument:
```

The second `onPageSizeChange` function is invoked when a user interacts with the "size selector" and so can be used to respond to a "page size" change (eg. updating the number of items listed in the page, updating the routing/URL, and/or updating other elements in the page).

### Routing/URL updates

If you want the pagination to change the URL of the page directly (eg. updating the query parameters) you need to pass the routing parameters to the component:

```handlebars
<Hds::Pagination::Numbered
  @totalItems={{this.demoTotalItems}}
  @itemsPerPage={{this.demoCurrentPageSize}}
  @currentPage={{this.demoCurrentPage}}
  @pageSizes={{this.demoPageSizes}}
  @route={{this.demoRouteName}}
  @queryFunction={{this.demoQueryFunctionNumbered}}
/>
```

where the `@queryFunction` function will be something like this:

```javascript
get demoQueryFunctionNumbered() {
  return (page, pageSize) => {
    return {
      demoCurrentPage: page,
      demoCurrentPageSize: pageSize,
      demoExtraParam: 'hello',
    };
  };
}
```

> When the routing parameters are provided, the "navigation controls" are rendered as links and if the user clicks on one of them the page URL is automatically updated. This means that the component's state is persisted **outside** of the component and so its whole state **must** be "controlled" by the consumer's code (otherwise there would be confliting states).

Reviewer: ```suggestion
When the routing parameters are provided, the "navigation controls" are rendered as links and if the user clicks on one of them the page URL is automatically updated. This means that the component’s state is persisted **outside** of the component and so its whole state **must** be "controlled" by the consumer’s code (otherwise there would be conflicting states).
```

> In this case, the component doesn't update its internal "state" but the value of the state variables (eg. `currentPage/currentPageSize`) is **always** determined by the consumer's controller code via the component's arguments (most of the times, they are directly connected to the query parameters in the URL).

Reviewer: ```suggestion
In this case, the component doesn’t update its internal "state" but the value of the state variables (eg. `currentPage/currentPageSize`) is **always** determined by the consumer’s controller code via the component’s arguments (most of the time, they are directly connected to the query parameters in the URL).
```

> The reason for using a consumer-side function to determine the `query` argument is because it's dynamic (it depends on the value of the `page` variable) and gives consumers the ability to specify their own query parameters (they will likely be different case by case).

Reviewer: ```suggestion
The reason for using a consumer-side function to determine the `query` argument is because it’s dynamic (it depends on the value of the `page` variable) and gives consumers the ability to specify their own query parameters (which will likely be different case by case).
```

> Even if the pagination is based on routing, the `onPageChange/onPageSizeChange` callbacks are still available and can be used to respond to the users' actions (eg. for logging, tracking, etc.)

Reviewer: ```suggestion
Even if the pagination is based on routing, the `onPageChange/onPageSizeChange` callbacks are still available and can be used to respond to the users’ actions (eg. for logging, tracking, etc.)
```

## How to use the "Compact" pagination

By default, the basic use of the pagination provides:

> The basic invocation of the "compact" pagination doesn't require any arguments (apart from the event/routing handlers, see below):

Reviewer: ```suggestion
The basic invocation of the "compact" pagination doesn’t require any arguments (apart from the event/routing handlers, see below):
```

```handlebars
<Hds::Pagination::Compact />
```

Renders to:

> In this variant only the "prev" and "next" navigation controls are displayed.

Reviewer: ```suggestion
In this variant, only the "prev" and "next" navigation controls are displayed.
```

### Extra arguments

> If necessary, it's possible to hide the labels in the controls:

Reviewer: ```suggestion
If necessary, it’s possible to hide the control labels:
```

```handlebars
<Hds::Pagination::Compact @showLabels={{false}} />
```

### Events handling

The component exposes a callback function that can be used to respond to page changes:

```handlebars
<Hds::Pagination::Compact
  @onPageChange={{this.handlePageChange}}
/>
```

The `onPageChange` function is invoked when a user interacts with a "prev" or "next" navigation control and so can be used to respond to a "page" change (eg. updating the list of items in the page and/or updating the routing/URL).

### Routing/URL updates

If you want the pagination to change the URL of the page directly (eg. updating the query parameters) you need to pass the routing parameters to the component:

```handlebars
<Hds::Pagination::Compact
  @route={{this.demoRouteName}}
  @queryFunction={{this.demoQueryFunctionCompact}}
  @isDisabledPrev={{this.demoIsDisabledPrev}}
  @isDisabledNext={{this.demoIsDisabledNext}}
/>
```

where the `@queryFunction` function will be something like this:

```javascript
get demoQueryFunctionCompact() {
  return (page) => {
    return {
      demoCurrentToken: page === 'prev' ? this.getPrevToken : this.getNextToken,
      demoExtraParam: 'hello',
    };
  };
}
```

> When the routing parameters are provided, the "prev/next" controls are rendered as links and if the user clicks on one of them the page URL is automatically updated. This means that the component's state is persisted **outside** of the component and so its whole state **must** be "controlled" by the consumer's code (otherwise there would be confliting states).

Reviewer: ```suggestion
When the routing parameters are provided, the "prev/next" controls are rendered as links and the page URL is automatically updated when the user clicks them. This means that the component’s state is persisted **outside** of the component and so its whole state **must** be "controlled" by the consumer’s code (otherwise there would be conflicting states).
```

> In this case, the component doesn't update its internal "state" but the value of the state variables is **always** determined by the consumer's controller code via the component's arguments (most of the times, they are directly connected to the query parameters in the URL).

Reviewer: ```suggestion
In this case, the component doesn’t update its internal "state" but the value of the state variables is **always** determined by the consumer’s controller code via the component’s arguments (usually, they are directly connected to the query parameters in the URL).
```

> The reason for using a consumer-side function to determine the `query` argument is because it's dynamic (it depends on the value of the `page/cursor` variable) and gives consumers the ability to specify their own query parameters (they will likely be different case by case).

Reviewer: ```suggestion
The reason for using a consumer-side function to determine the `query` argument is because it’s dynamic (it depends on the value of the `page/cursor` variable) and gives consumers the ability to specify their own query parameters (they will likely be different case by case).
```
